### PR TITLE
Fixed assembly error with hangfire queues. Added hosted service to register recurring jobs automatically. Updated pending submissions job result. 

### PR DESCRIPTION
**Pull request by shsimeonova**
Closes https://github.com/SoftUni-Internal/exam-systems-issues/issues/892

- Created HangfireJobsBusinessService and moved interface into common. By doing this the assembly error is resolved.
- Updated registration of hangfire application servers by adding queues (application name for now) so that every server picks up its own background jobs
- Updated result of enqueue pending submissions job to return count of submissions
- Added hosted service implemetation in Administration that registers reccuring jobs on app start

### CODE REVIEW

File: Servers/UI/OJS.Servers.Ui.Infrastructure/Extensions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -6,7 +6,9 @@ namespace OJS.Servers.Ui.Infrastructure.Extensions
     using OJS.Data;
     using OJS.Data.Models.Users;
     using OJS.Servers.Infrastructure.Extensions;
+    using OJS.Services.Common;
     using OJS.Services.Common.Models.Configurations;
+    using OJS.Services.Ui.Business.Implementations;
```
Comment (gogo4ds) [inline Servers/UI/OJS.Servers.Ui.Infrastructure/Extensions/ServiceCollectionExtensions.cs:11]: Not used

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs [hunk 2]
```diff
@@ -1,5 +1,6 @@
 ﻿namespace OJS.Servers.Ui.Controllers.Api;
 
+using OJS.Services.Common;
```
Comment (gogo4ds) [inline Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs:3]: Not used

File: Services/Administration/OJS.Services.Administration.Business/OJS.Services.Administration.Business.csproj [hunk 3]
```diff
@@ -14,6 +14,7 @@
 
     <ItemGroup>
       <PackageReference Include="AutoCrudAdmin" Version="1.1.26" />
+      <PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" Version="6.0.0" />
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/OJS.Services.Administration.Business.csproj:17]: Not used

File: Services/Common/OJS.Services.Common/BackgroundJobsHostedService.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                m => m.DeleteProcessedSubmissionsJob(),
+                BackgroundJobs.JobNames.DeleteOldSubmissionsJobCron,
+                this.applicationQueueName);
+
+        this.hangfireBackgroundJobs
+            .AddOrUpdateRecurringJob<IHangfireBackgroundJobsBusinessService>(
+                BackgroundJobs.JobNames.DeleteOldSubmissionsJobName,
+                m => m.DeleteProcessedSubmissionsJob(),
+                BackgroundJobs.JobNames.DeleteOldSubmissionsJobCron,
+                this.applicationQueueName);
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common/BackgroundJobsHostedService.cs:75]: This is the same as the above registration.

File: Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+    public interface IHangfireBackgroundJobsBusinessService : IService
+    {
+        Task ResetAllProcessingSubmissionsJob();
+
+        int EnqueuePendingSubmissionsJob();
+
+        void DeleteProcessedSubmissionsJob();
+
+        Task<int> GetUnprocessedTotalCountJob();
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs:14]: Why is this a background job?

File: Services/Common/OJS.Services.Common/ServiceConstants.cs [hunk 6]
```diff
@@ -95,5 +95,18 @@ public static class ExecutionStrategyNames
...
             }
         }
+
+        public static class BackgroundJobs
+        {
+            public static class JobNames
+            {
+                public const string EnqueuePendingSubmissionsJobName = "EnqueuePendingSubmissionsForProcessing";
+                public const string EnqueuePendingSubmissionsJobCron = "*/3 * * * *";
+                public const string EnqueuePendingSubmissionsJobAddedMessage = $"{EnqueuePendingSubmissionsJobName} job added successfully! It will run at {EnqueuePendingSubmissionsJobCron}";
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common/ServiceConstants.cs:105]: These are 3 different types of information. Maybe split in different classes or just put in class BackgroundJobs, but not all in JobNames

File: Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+        if (allProcessingSubmissionIds.Count() <= 0)
+        {
+            return;
+        }
+
+        foreach (var submissionForProcessingId in allProcessingSubmissionIds)
+        {
+            await this.submissionsForProcessingData.ResetProcessingStatusById(submissionForProcessingId);
+        }
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs:38]: This logic is different from the one in Administration. Maybe UI should not execute this job

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Common/OJS.Services.Common/BackgroundJobsHostedService.cs (hunk 4 above)
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common/BackgroundJobsHostedService.cs]: Removed

File: Services/Administration/OJS.Services.Administration.Business/OJS.Services.Administration.Business.csproj (hunk 3 above)
Comment (shsimeonova) [reply Services/Administration/OJS.Services.Administration.Business/OJS.Services.Administration.Business.csproj]: Removed

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs (hunk 2 above)
Comment (shsimeonova) [reply Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs:3]: Removed

File: Servers/UI/OJS.Servers.Ui.Infrastructure/Extensions/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (shsimeonova) [reply Servers/UI/OJS.Servers.Ui.Infrastructure/Extensions/ServiceCollectionExtensions.cs]: Removed

File: Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs (hunk 5 above)
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs]: Copy-paste mistake.Removed this.

File: Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs (hunk 7 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs]: Removed.

File: Services/Common/OJS.Services.Common/ServiceConstants.cs (hunk 6 above)
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common/ServiceConstants.cs]: Moved them in separate classes

File: Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿namespace OJS.Services.Common
+{
+    using SoftUni.Services.Infrastructure;
+
+    public interface IHangfireBackgroundJobsBusinessService : IService
+    {
+        int EnqueuePendingSubmissionsJob();
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs:7]: No need to put "Job" at the end

File: Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+    public HangfireBackgroundJobsBusinessService(
+        ISubmissionsForProcessingCommonDataService submissionsForProcessingData,
+        IUserProviderService userProviderService)
+    {
+        this.submissionsForProcessingData = submissionsForProcessingData;
+        this.userProviderService = userProviderService;
+    }
+
+    public Task<int> GetUnprocessedTotalCountJob()
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs:20]: If this is not part of the interface, should not be here as well

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Administration/OJS.Services.Administration.Business/Implementations/HangfireBackgroundJobsBusinessService.cs [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+    public class HangfireBackgroundJobsBusinessService : IHangfireBackgroundJobsBusinessService
+    {
+        private readonly ISubmissionsForProcessingBusinessService submissionsForProcessing;
+
+        public HangfireBackgroundJobsBusinessService(ISubmissionsForProcessingBusinessService submissionsForProcessing)
+            => this.submissionsForProcessing = submissionsForProcessing;
+
+        public int EnqueuePendingSubmissionsJob() => this.submissionsForProcessing.EnqueuePendingSubmissions();
+
+        public void DeleteProcessedSubmissionsJob() => throw new NotImplementedException();
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/Implementations/HangfireBackgroundJobsBusinessService.cs:16]: This is not implemented anywhere

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Administration/OJS.Services.Administration.Business/Implementations/HangfireBackgroundJobsBusinessService.cs (hunk 10 above)
Comment (shsimeonova) [reply Services/Administration/OJS.Services.Administration.Business/Implementations/HangfireBackgroundJobsBusinessService.cs:16]: This is a mistake, It should be implemented here. I'll fix it.

File: Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs (hunk 9 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs]: Removed

File: Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs (hunk 8 above)
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common/IHangfireBackgroundJobsBusinessService.cs]: Renamed methods

File: Services/Common/OJS.Services.Common/BackgroundJobsHostedService.cs [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using static ServiceConstants;
+
+public class BackgroundJobsHostedService : IHostedService
+{
+    private readonly IHangfireBackgroundJobsService hangfireBackgroundJobs;
+    private readonly ILogger<BackgroundJobsHostedService> logger;
+
+    private readonly string applicationQueueName = ApplicationName.Administration.ToString();
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common/BackgroundJobsHostedService.cs:18]: As this is common now, this should be `administrationQueueName`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+public class HangfireBackgroundJobsBusinessService : IHangfireBackgroundJobsBusinessService
+{
+    public HangfireBackgroundJobsBusinessService()
+    {
+    }
+
+    public int EnqueuePendingSubmissions() => throw new System.NotImplementedException();
+
+    public void DeleteProcessedSubmissions() => throw new System.NotImplementedException();
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/HangfireBackgroundJobsBusinessService.cs:13]: Write comments on why this is not implemented.

**Review by gogo4ds — APPROVED** [review]

**Review by gogo4ds — APPROVED** [review]

**Review by gogo4ds — DISMISSED** [review]

**Review by gogo4ds — DISMISSED** [review]

**Review by gogo4ds — APPROVED** [review]

**Pull request merged**